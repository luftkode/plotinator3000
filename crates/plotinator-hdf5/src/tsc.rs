use std::{path::Path, time::Instant};

use anyhow::bail;
use chrono::{DateTime, Utc};
use plotinator_log_if::{
    hdf5::SkytemHdf5,
    prelude::{PlotLabels, Plotable},
    rawplot::RawPlot,
};
use serde::{Deserialize, Serialize};

use crate::tsc::{
    gps_marks::GpsMarkRecords, gps_pvt::GpsPvtRecords, hm::HmData, metadata::RootMetadata,
};

mod gps_marks;
mod gps_pvt;
mod hm;
mod metadata;

pub(crate) const TSC_LEGEND_NAME: &str = "TSC";

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Tsc {
    first_timestamp: DateTime<Utc>,
    raw_plots: Vec<RawPlot>,
    metadata: Vec<(String, String)>,
}

impl Tsc {
    const DESCRIPTIVE_NAME: &str = "TSC";
}

impl SkytemHdf5 for Tsc {
    fn from_path(path: impl AsRef<Path>) -> anyhow::Result<Self> {
        let total_start = Instant::now();
        let start_reading = Instant::now();
        let h5 = hdf5::File::open(path)?;
        log::info!("== Reading TSC datasets");

        // Read top level metadata
        let root_metadata = RootMetadata::parse_from_tsc(&h5)?;
        let mut metadata = root_metadata.metadata_strings();

        let start = Instant::now();
        let hm = HmData::from_hdf5(&h5)?;
        log::info!("Read HmData in {:.1?}", start.elapsed());

        let start = Instant::now();
        let gps_marks = GpsMarkRecords::from_hdf5(&h5)?;
        log::info!("Read GpsMarks in {:.1?}", start.elapsed());
        let start = Instant::now();
        let gps_pvts = GpsPvtRecords::from_hdf5(&h5)?;
        log::info!("Read GpsPvt in {:.1?}", start.elapsed());
        log::info!(
            "== Finished reading datasets in {:.1?}",
            start_reading.elapsed()
        );

        let Some(first_timestamp) = gps_marks.first_timestamp() else {
            bail!("gps marks dataset is empty");
        };

        let start_building_plots = Instant::now();
        log::info!("== Building plots");
        let start = Instant::now();
        log::info!("Creating GPS marks plots");
        let (mut plots, mut gps_metadata) = gps_marks.build_plots_and_metadata();
        metadata.append(&mut gps_metadata);
        log::info!("Created GPS marks plots in {:.1?}", start.elapsed());
        log::info!("Creating GPS PVT plots");
        let start = Instant::now();
        plots.extend(gps_pvts.build_plots());
        log::info!("Created GPS PVT plots in {:.1?}", start.elapsed());
        log::info!("Creating HM plots");
        let start = Instant::now();
        let gps_time = gps_marks.timestamps();
        let (hm_plots, mut hm_metadata) = hm.build_plots_and_metadata(&gps_time, &root_metadata)?;
        plots.extend(hm_plots);
        metadata.append(&mut hm_metadata);
        log::info!("Created HM plots in {:.1?}", start.elapsed());
        log::info!(
            "== Finishing build plots in {:.1?}",
            start_building_plots.elapsed()
        );
        log::info!(
            "== == Total TSC::from_path duration: {:.1?} == ==",
            total_start.elapsed()
        );

        plots.retain(|p| match p {
            RawPlot::GeoSpatialDataset(_) => true,
            RawPlot::Generic { common } | RawPlot::Boolean { common } => {
                if common.points().len() < 2 {
                    log::warn!("Discarding plot with less than 2 points: {}", common.name());
                    false
                } else {
                    true
                }
            }
        });

        Ok(Self {
            first_timestamp,
            raw_plots: plots,
            metadata,
        })
    }
}

impl Plotable for Tsc {
    fn raw_plots(&self) -> &[RawPlot] {
        &self.raw_plots
    }

    fn first_timestamp(&self) -> DateTime<Utc> {
        self.first_timestamp
    }

    fn descriptive_name(&self) -> &str {
        Self::DESCRIPTIVE_NAME
    }

    fn labels(&self) -> Option<&[PlotLabels]> {
        None
    }

    fn metadata(&self) -> Option<Vec<(String, String)>> {
        Some(self.metadata.clone())
    }
}

#[cfg(test)]
mod tests {
    use crate::tsc::{gps_marks::GpsMarkRecords, hm::HmData, metadata::RootMetadata};
    use plotinator_log_if::rawplot::RawPlot;
    use plotinator_test_util::test_file_defs::tsc::*;
    use testresult::TestResult;

    #[test]
    fn test_read_tsc_gps_marks_timestamps() -> TestResult {
        let h5file = hdf5::File::open(tsc())?;

        let gps_marks = GpsMarkRecords::from_hdf5(&h5file)?;
        let gps_time = gps_marks.timestamps();
        let first_10_gps_time: Vec<f64> = gps_time.into_iter().take(10).collect();
        insta::assert_debug_snapshot!(first_10_gps_time);

        Ok(())
    }

    #[test]
    fn test_read_bfield_and_zero_pos() -> TestResult {
        let h5file = hdf5::File::open(tsc())?;
        let root_metadata = RootMetadata::parse_from_tsc(&h5file)?;
        let gps_marks = GpsMarkRecords::from_hdf5(&h5file)?;
        let hm_data = HmData::from_hdf5(&h5file)?;

        let gps_timestamps = gps_marks.timestamps();
        let (plots, _metadata) =
            hm_data.build_plots_and_metadata(&gps_timestamps, &root_metadata)?;

        let zero_pos_first_10: Vec<[f64; 2]> = match plots.first().unwrap() {
            RawPlot::Generic { common } => common.points().iter().copied().take(10).collect(),
            _ => unreachable!(),
        };
        let bfield_first_10: Vec<[f64; 2]> = match &plots[1] {
            RawPlot::Generic { common } => common.points().iter().copied().take(10).collect(),
            _ => unreachable!(),
        };

        insta::assert_debug_snapshot!(zero_pos_first_10);
        insta::assert_debug_snapshot!(bfield_first_10);

        Ok(())
    }
}
