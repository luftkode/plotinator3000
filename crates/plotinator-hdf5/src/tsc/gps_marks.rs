use chrono::{DateTime, TimeZone as _, Utc};
use hdf5::H5Type;
use ndarray::{ArrayBase, Dim, OwnedRepr};
use plotinator_log_if::{
    leap_seconds::{GpsWeek, TowMs, TowSubMs, gps_to_unix_ns},
    prelude::{ExpectedPlotRange, RawPlotCommon},
    rawplot::RawPlot,
};

type GpsMarks = ArrayBase<OwnedRepr<GpsMarkRecord>, Dim<[usize; 1]>>;

/// Wrapper around all the [`GpsMarkRecord`]s from a TSC.h5 file
pub(crate) struct GpsMarkRecords {
    inner: GpsMarks,
}

impl GpsMarkRecords {
    const DATASET_NAME: &str = "GPS_marks";

    pub fn from_hdf5(h5: &hdf5::File) -> hdf5::Result<Self> {
        let dataset = h5.dataset(Self::DATASET_NAME)?;
        let gps_marks = dataset.read::<GpsMarkRecord, ndarray::Ix1>()?;

        Ok(Self { inner: gps_marks })
    }

    // Return a vector of timestamps (unix UTC nanoseconds)
    pub fn timestamps(&self) -> Vec<f64> {
        self.inner
            .iter()
            .map(|g| g.unix_timestamp_ns() as f64)
            .collect()
    }

    pub fn first_timestamp(&self) -> Option<DateTime<Utc>> {
        self.inner.first().map(|e| e.unix_timestamp_utc())
    }

    #[allow(clippy::too_many_lines, reason = "Long but simple")]
    pub fn build_plots_and_metadata(&self) -> (Vec<RawPlot>, Vec<(String, String)>) {
        let time = self.timestamps();

        let mut metadata = vec![("GPS Marks".to_owned(), time.len().to_string())];

        let mut pulse_width = Vec::with_capacity(time.len());
        let mut acc_est = Vec::with_capacity(time.len());
        let mut count = Vec::with_capacity(time.len());

        // flag series
        let mut mode_running = Vec::with_capacity(time.len());
        let mut run = Vec::with_capacity(time.len());
        let mut new_falling_edge = Vec::with_capacity(time.len());
        let mut timebase_gnss = Vec::with_capacity(time.len());
        let mut timebase_utc = Vec::with_capacity(time.len());
        let mut utc_avail = Vec::with_capacity(time.len());
        let mut time_valid = Vec::with_capacity(time.len());
        let mut new_rising_edge = Vec::with_capacity(time.len());

        let mut step_counter = StepCounter::new();
        let mut delta_computer = DeltaComputer::new();

        for (e, &t) in self.inner.iter().zip(&time) {
            step_counter.record_count(e.count, t);
            delta_computer.record_timestamp(e.unix_timestamp_ns(), t, e.count);

            pulse_width.push([t, e.pulse_width_us()]);
            acc_est.push([t, e.acc_est as f64]);
            count.push([t, e.count as f64]);

            mode_running.push([t, e.flag(GpsMarkRecord::FLAGS_MODE_RUNNING)]);
            run.push([t, e.flag(GpsMarkRecord::FLAGS_RUN)]);
            new_falling_edge.push([t, e.flag(GpsMarkRecord::FLAGS_NEWFALLINGEDGE)]);
            timebase_gnss.push([t, e.flag(GpsMarkRecord::FLAGS_TIMEBASE_GNSS)]);
            timebase_utc.push([t, e.flag(GpsMarkRecord::FLAGS_TIMEBASE_UTC)]);
            utc_avail.push([t, e.flag(GpsMarkRecord::FLAGS_UTC_AVAIL)]);
            time_valid.push([t, e.flag(GpsMarkRecord::FLAGS_TIME_VALID)]);
            new_rising_edge.push([t, e.flag(GpsMarkRecord::FLAGS_NEWRISINGEDGE)]);
        }

        // Add step counter metadata
        metadata.push((
            "Count step differences".to_owned(),
            step_counter.to_metadata_string(),
        ));
        let (all_stats, normal_stats) = delta_computer.to_metadata_string();
        metadata.push(("Timestamp Δt stats [s]".to_owned(), all_stats));
        metadata.push((
            "Timestamp Δt stats (without faults) [s]".to_owned(),
            normal_stats,
        ));
        (
            vec![
                RawPlotCommon::new(
                    "Pulse width [µs]".to_owned(),
                    pulse_width,
                    ExpectedPlotRange::Thousands,
                )
                .into(),
                RawPlotCommon::new(
                    "Accuracy estimate [ns]".to_owned(),
                    acc_est,
                    ExpectedPlotRange::OneToOneHundred,
                )
                .into(),
                RawPlotCommon::new("Count".to_owned(), count, ExpectedPlotRange::Thousands).into(),
                RawPlotCommon::new(
                    "Mode running [bool]".to_owned(),
                    mode_running,
                    ExpectedPlotRange::Percentage,
                )
                .into(),
                RawPlotCommon::new("Run [bool]".to_owned(), run, ExpectedPlotRange::Percentage)
                    .into(),
                RawPlotCommon::new(
                    "New falling edge [bool]".to_owned(),
                    new_falling_edge,
                    ExpectedPlotRange::Percentage,
                )
                .into(),
                RawPlotCommon::new(
                    "Timebase GNSS [bool]".to_owned(),
                    timebase_gnss,
                    ExpectedPlotRange::Percentage,
                )
                .into(),
                RawPlotCommon::new(
                    "Timebase UTC [bool]".to_owned(),
                    timebase_utc,
                    ExpectedPlotRange::Percentage,
                )
                .into(),
                RawPlotCommon::new(
                    "UTC available [bool]".to_owned(),
                    utc_avail,
                    ExpectedPlotRange::Percentage,
                )
                .into(),
                RawPlotCommon::new(
                    "Time valid [bool]".to_owned(),
                    time_valid,
                    ExpectedPlotRange::Percentage,
                )
                .into(),
                RawPlotCommon::new(
                    "New rising edge [bool]".to_owned(),
                    new_rising_edge,
                    ExpectedPlotRange::Percentage,
                )
                .into(),
                RawPlotCommon::new(
                    "Timestamp Δt [s]".to_owned(),
                    delta_computer.take_plot_points(),
                    ExpectedPlotRange::OneToOneHundred,
                )
                .into(),
            ],
            metadata,
        )
    }
}

/// Tracks step differences in GPS mark counts with detailed information
struct StepCounter {
    events: Vec<StepEvent>,
    prev_count: Option<u16>,
    prev_time: Option<f64>,
}

/// Represents a step event with detailed information
struct StepEvent {
    step_delta: u16,
    timestamp_ns: i64,
    time_delta_s: f64,
    count_before: u16,
    count_after: u16,
}

impl StepCounter {
    fn new() -> Self {
        Self {
            events: Vec::new(),
            prev_count: None,
            prev_time: None,
        }
    }

    fn record_count(&mut self, count: u16, timestamp_ns: f64) {
        if let (Some(prev_count), Some(prev_time_ns)) = (self.prev_count, self.prev_time) {
            let step_delta = count.wrapping_sub(prev_count);

            if step_delta != 1 {
                let time_delta_s = (timestamp_ns - prev_time_ns) / 1_000_000_000.0;
                self.events.push(StepEvent {
                    step_delta,
                    timestamp_ns: timestamp_ns as i64,
                    time_delta_s,
                    count_before: prev_count,
                    count_after: count,
                });
            }
        }

        self.prev_count = Some(count);
        self.prev_time = Some(timestamp_ns);
    }

    fn to_metadata_string(&self) -> String {
        if self.events.is_empty() {
            return "All steps were 1".to_owned();
        }

        let mut result = String::new();
        result.push_str(&format!(
            "Found {} non-1 step event{plural}:\n",
            self.events.len(),
            plural = if self.events.len() == 1 { "" } else { "s" }
        ));

        for event in &self.events {
            let dt = Utc.timestamp_nanos(event.timestamp_ns);
            let time_str = dt.format("%H:%M:%S%.3f").to_string();

            result.push_str(&format!(
                "Step={} at {time_str} [time delta: {:.6}s, count: {} -> {}]\n",
                event.step_delta, event.time_delta_s, event.count_before, event.count_after
            ));
        }

        result
    }
}

/// Computes deltas between consecutive GPS marks
struct DeltaComputer {
    deltas: Vec<[f64; 2]>, // [timestamp, delta_s]
    prev_ns: Option<i64>,
    all_stats: Vec<f64>,     // all deltas
    normal_stats: Vec<f64>,  // deltas without faults (Δcount = 1)
    prev_count: Option<u16>, // to detect faults
}

impl DeltaComputer {
    fn new() -> Self {
        Self {
            deltas: Vec::new(),
            prev_ns: None,
            all_stats: Vec::new(),
            normal_stats: Vec::new(),
            prev_count: None,
        }
    }

    fn record_timestamp(&mut self, curr_ns: i64, curr_time: f64, curr_count: u16) {
        if let Some(prev) = self.prev_ns {
            let delta_s = (curr_ns - prev) as f64 / 1e9;
            self.deltas.push([curr_time, delta_s]);
            self.all_stats.push(delta_s);

            // Add to normal_stats only if step increment is 1
            if let Some(prev_count) = self.prev_count {
                if curr_count.wrapping_sub(prev_count) == 1 {
                    self.normal_stats.push(delta_s);
                }
            }
        }

        self.prev_ns = Some(curr_ns);
        self.prev_count = Some(curr_count);
    }

    fn take_plot_points(self) -> Vec<[f64; 2]> {
        self.deltas
    }

    fn stats_to_string(stats: &[f64]) -> String {
        if stats.is_empty() {
            return "No deltas".to_owned();
        }
        let min = stats.iter().copied().fold(f64::INFINITY, f64::min);
        let max = stats.iter().copied().fold(f64::NEG_INFINITY, f64::max);
        let avg = stats.iter().sum::<f64>() / stats.len() as f64;
        let mean_diff_sq_sum: f64 = stats.iter().map(|x| (x - avg).powi(2)).sum();
        let stddev = (mean_diff_sq_sum / stats.len() as f64).sqrt();
        format!("min={min:.6}, max={max:.6}, avg={avg:.6}, σ={stddev:.6}")
    }

    fn to_metadata_string(&self) -> (String, String) {
        let all = Self::stats_to_string(&self.all_stats);
        let normal = Self::stats_to_string(&self.normal_stats);
        (all, normal)
    }
}

/// TIM-TM2 (0x0D, 0x03)
/// Time mark data
///
/// Description for details.
///
/// Supported on:
///  - u-blox 8 / u-blox M8 with protocol version 22 (only with Timing Products)
///
/// `CLASS_ID` = 13, `MESSAGE_ID` = 3
///
/// Flag constants:
/// - `FLAGS_MODE_RUNNING` = 1     // single = 0, running = 1
/// - `FLAGS_RUN` = 2              // armed = 0, stopped = 1
/// - `FLAGS_NEWFALLINGEDGE` = 4   // new falling edge detected
/// - `FLAGS_TIMEBASE_GNSS` = 8    // 0 = time base is receiver time, 1 = time base is GNSS Time (according to the configuration in CFG-TP5 for tpldx= 0)
/// - `FLAGS_TIMEBASE_UTC` = 16    // Time Base is UTC (the variant according to the configuration in CFG-NAV5
/// - `FLAGS_UTC_AVAIL` = 32       // 0 = utc not available, 1 = utc available
/// - `FLAGS_TIME_VALID` = 64      // 0 = time is not valid, 1 time is valid
/// - `FLAGS_NEWRISINGEDGE` = 128  // new rising edge detected
#[derive(H5Type, Debug)]
#[repr(C)]
struct GpsMarkRecord {
    /// Channel (i.e. EXTINT) upon which the pulse was measured
    ch: u8,

    /// Bitmask [newRisingEdge, time, utc, timeBase, , newFallingEdge, run, mode]
    flags: u8,

    /// Rising edge count
    count: u16,

    /// Week number of last rising edge
    #[hdf5(rename = "wnR")]
    wn_r: u16,

    /// Week number of last falling edge
    #[hdf5(rename = "wnF")]
    wn_f: u16,

    /// Time of Week of rising edge (milliseconds)
    #[hdf5(rename = "towMsR")]
    tow_ms_r: u32,

    /// Millisecond fraction of `ToW` of rising edge in nanoseconds
    #[hdf5(rename = "towSubMsR")]
    tow_sub_ms_r: u32,

    /// Time of Week of falling edge (milliseconds)
    #[hdf5(rename = "towMsF")]
    tow_ms_f: u32,

    /// Millisecond fraction of `ToW` of falling edge in nanoseconds
    #[hdf5(rename = "towSubMsF")]
    tow_sub_ms_f: u32,

    /// Accuracy estimate
    #[hdf5(rename = "accEst")]
    acc_est: u32,
}

impl GpsMarkRecord {
    const FLAGS_MODE_RUNNING: u8 = 1;
    const FLAGS_RUN: u8 = 2;
    const FLAGS_NEWFALLINGEDGE: u8 = 4;
    const FLAGS_TIMEBASE_GNSS: u8 = 8;
    const FLAGS_TIMEBASE_UTC: u8 = 16;
    const FLAGS_UTC_AVAIL: u8 = 32;
    const FLAGS_TIME_VALID: u8 = 64;
    const FLAGS_NEWRISINGEDGE: u8 = 128;

    fn flag(&self, mask: u8) -> f64 {
        if self.flags & mask != 0 { 1.0 } else { 0.0 }
    }

    fn unix_timestamp_utc(&self) -> DateTime<Utc> {
        Utc.timestamp_nanos(self.unix_timestamp_ns())
    }

    /// Convert GPS week + TOW to precise Unix timestamp in nanoseconds since Unix epoch
    /// All calculations done in integers to avoid floating point accumulation errors
    fn unix_timestamp_ns(&self) -> i64 {
        gps_to_unix_ns(
            GpsWeek(self.wn_r),
            TowMs(self.tow_ms_r),
            TowSubMs(self.tow_sub_ms_r),
        )
        .0
    }

    /// Calculate pulse width in microseconds (integer arithmetic until final conversion)
    pub fn pulse_width_us(&self) -> f64 {
        const MS_TO_NS: i64 = 1_000_000;
        const NS_TO_US: f64 = 1.0 / 1_000.0;

        // Compute deltas in each unit
        let week_delta = (self.wn_f as i64 - self.wn_r as i64) * 604_800; // seconds
        let ms_delta = self.tow_ms_f as i64 - self.tow_ms_r as i64; // milliseconds
        let sub_ns_delta = self.tow_sub_ms_f as i64 - self.tow_sub_ms_r as i64; // nanoseconds

        // Combine everything into nanoseconds
        let total_ns = week_delta * 1_000_000_000   // weeks -> ns
                    + ms_delta * MS_TO_NS // ms -> ns
                    + sub_ns_delta; // ns

        // Convert to microseconds
        total_ns as f64 * NS_TO_US
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::NaiveDate;
    use plotinator_test_util::test_file_defs::tsc::*;
    use testresult::TestResult;

    #[test]
    fn read_gps_marks() -> TestResult {
        let h5file = hdf5::File::open(tsc())?;
        let gps_marks = GpsMarkRecords::from_hdf5(&h5file)?;
        let gps_marks = gps_marks.inner;
        for g in &gps_marks {
            eprintln!("g: {g:?}");
            eprintln!("{}", g.unix_timestamp_utc());
        }

        insta::assert_debug_snapshot!(gps_marks);
        Ok(())
    }

    #[test]
    fn test_timestamp_conversion_simple() {
        let record = GpsMarkRecord {
            ch: 1,
            flags: 0,
            count: 100,
            wn_r: 2200, // Example GPS week
            wn_f: 0,
            tow_ms_r: 123_000, // 123 s
            tow_sub_ms_r: 0,   // 0 nanoseconds
            tow_ms_f: 0,
            tow_sub_ms_f: 0,
            acc_est: 0,
        };

        let dt = Utc.timestamp_opt(1646524905, 0).unwrap();
        let readable_dt = NaiveDate::from_ymd_opt(2022, 3, 6) // YYYY-MM-DD
            .unwrap()
            .and_hms_nano_opt(0, 1, 45, 0) // HH:MM:SS + nanos
            .unwrap()
            .and_utc();
        assert_eq!(dt, readable_dt, "Reference timestamps don't match");
        assert_eq!(
            record.unix_timestamp_utc(),
            readable_dt,
            "invalid conversion"
        );
    }

    #[test]
    fn test_timestamp_conversion_nanos() {
        const NANO_SEC_PART: u32 = 814_123_456;
        let record = GpsMarkRecord {
            ch: 1,
            flags: 0,
            count: 100,
            wn_r: 2381, // Example GPS week
            wn_f: 2381,
            tow_ms_r: 62_625_000,        // 62625000 ms
            tow_sub_ms_r: NANO_SEC_PART, // 814_123_456 nanoseconds
            tow_ms_f: 0,
            tow_sub_ms_f: 0,
            acc_est: 0,
        };

        let dt = Utc.timestamp_nanos(1_756_056_207_814_123_456);
        let readable_dt = NaiveDate::from_ymd_opt(2025, 8, 24) // YYYY-MM-DD
            .unwrap()
            .and_hms_nano_opt(17, 23, 27, NANO_SEC_PART) // HH:MM:SS + nanos
            .unwrap()
            .and_utc();
        assert_eq!(dt, readable_dt, "Reference timestamps don't match");
        assert_eq!(
            record.unix_timestamp_utc(),
            readable_dt,
            "invalid conversion"
        );
    }
}
