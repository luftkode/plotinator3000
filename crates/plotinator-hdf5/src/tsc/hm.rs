use anyhow::{bail, ensure};
use ndarray::{ArrayBase, Dim, OwnedRepr};
use plotinator_log_if::prelude::RawPlot;

// 6-dimensional array: [N, 413, 2, 76, 6, 4]
type HmArray = ArrayBase<OwnedRepr<i64>, Dim<[usize; 6]>>;

/// Wrapper around the hm dataset from `/RX/monomial_basis_data/hm`
pub(crate) struct HmData<'h5> {
    h5file: &'h5 hdf5::File, // Keep file open for future access
    dataset_name: String,    // Path to dataset
    shape: [usize; 6],       // Only load shape initially
    inner: Option<HmArray>,  // Load full data lazily when needed
}

impl<'h5> HmData<'h5> {
    pub fn from_hdf5(h5: &'h5 hdf5::File) -> hdf5::Result<Self> {
        let dataset_name = "/RX/monomial_basis_data/hm".to_owned();
        let dataset = h5.dataset(&dataset_name)?;
        let shape_vec = dataset.shape();
        let shape: [usize; 6] = shape_vec
            .try_into()
            .map_err(|e| hdf5::Error::Internal(format!("Dataset is not 6D: {e:?}")))?;

        Ok(Self {
            h5file: h5,
            dataset_name,
            shape,
            inner: None, // don't read data yet
        })
    }

    pub fn shape(&self) -> &[usize] {
        self.shape.as_ref()
    }

    /// Load full data only if needed
    #[allow(
        dead_code,
        reason = "We will need this later, and want to test that the functionality doesn't break"
    )]
    pub fn load_full(&mut self) -> hdf5::Result<()> {
        if self.inner.is_none() {
            let dataset = self.h5file.dataset(&self.dataset_name)?;
            let hm_data = dataset.read::<i64, _>()?;
            self.inner = Some(hm_data);
        }
        Ok(())
    }

    #[allow(
        dead_code,
        reason = "We will need this later, and want to test that the functionality doesn't break"
    )]
    /// Create time-aligned point series from GPS timestamps
    /// Returns Vec<[f64; 2]> where each element is [timestamp, `hm_value`]
    /// The first dimension of hm data should match (or be 1 less than) GPS marks count
    pub fn create_time_series(
        &self,
        gps_timestamps: &[f64],
        coords: [usize; 5], // [dim1, dim2, dim3, dim4, dim5] - first dim varies with time
    ) -> anyhow::Result<Vec<[f64; 2]>> {
        let hm_len = self.shape()[0]; // First dimension length
        let gps_len = gps_timestamps.len();

        // Validate alignment: hm length should match GPS or be 1 less
        ensure!(
            hm_len == gps_len || hm_len == gps_len - 1,
            format!(
                "HM data length ({hm_len}) doesn't align with GPS timestamps ({gps_len}). Expected {gps_len} or {}",
                gps_len - 1
            )
        );

        let mut time_series = Vec::with_capacity(hm_len);
        let [dim1, dim2, dim3, dim4, dim5] = coords;

        for (i, gps_ts) in gps_timestamps.iter().enumerate().take(hm_len) {
            if let Some(hm_value) = self.get([i, dim1, dim2, dim3, dim4, dim5]) {
                time_series.push([*gps_ts, *hm_value as f64]);
            } else {
                bail!(
                    "Invalid HM coordinates at index {i}: [{i}, {dim1}, {dim2}, {dim3}, {dim4}, {dim5}]"
                );
            }
        }

        Ok(time_series)
    }

    #[allow(
        dead_code,
        reason = "We will need this later, and want to test that the functionality doesn't break"
    )]
    /// Access element at specific coordinates [dim0, dim1, dim2, dim3, dim4, dim5]
    /// Shape is [N, 413, 2, 76, 6, 4]
    pub fn get(&self, coords: [usize; 6]) -> Option<&i64> {
        self.inner.as_ref()?.get(coords)
    }

    #[allow(
        dead_code,
        reason = "We will need this later, and want to test that the functionality doesn't break"
    )]
    /// Get a slice along the first dimension at fixed other coordinates
    /// Example: `get_slice_dim0([0, 1, 1, 1, 1])`
    pub fn get_slice_dim0(&self, coords: [usize; 5]) -> Option<ndarray::ArrayView1<i64>> {
        let [dim1, dim2, dim3, dim4, dim5] = coords;
        self.inner
            .as_ref()?
            .slice(ndarray::s![.., dim1, dim2, dim3, dim4, dim5])
            .into_dimensionality()
            .ok()
    }

    // Build metadata without loading full data
    pub fn build_plots_and_metadata(
        &self,
        gps_timestamps: &[f64],
    ) -> (Vec<RawPlot>, Vec<(String, String)>) {
        let hm_len = self.shape()[0];
        let gps_len = gps_timestamps.len();
        let metadata = vec![
            ("HM length".to_owned(), hm_len.to_string()),
            ("GPS length".to_owned(), gps_len.to_string()),
        ];
        (vec![], metadata)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use plotinator_test_util::test_file_defs::tsc::*;
    use testresult::TestResult;

    #[test]
    fn read_hm_data() -> TestResult {
        let h5file = hdf5::File::open(tsc())?;
        let mut hm_data = HmData::from_hdf5(&h5file)?;

        println!("HM Data shape: {:?}", hm_data.shape());

        // Test accessing specific elements (triggers lazy load)
        hm_data.load_full()?; // optional, can rely on get() lazy load
        if let Some(value) = hm_data.get([0, 0, 0, 0, 0, 0]) {
            println!("Element [0,0,0,0,0,0]: {value}");
        }

        // Test accessing a slice
        if let Some(slice) = hm_data.get_slice_dim0([0, 0, 0, 0, 0]) {
            println!("Slice along first dimension: {slice:?}");
        }

        // Take a small sample for snapshot testing to avoid huge output
        let sample_coords = [
            [0, 0, 0, 0, 0, 0],
            [0, 1, 0, 0, 0, 0],
            [1, 0, 1, 0, 0, 0],
            [2, 10, 0, 10, 0, 0],
            [3, 100, 1, 50, 5, 3],
        ];

        let mut sample_values = Vec::new();
        for coords in sample_coords {
            if let Some(value) = hm_data.get(coords) {
                sample_values.push((coords, *value));
            }
        }

        insta::assert_debug_snapshot!(sample_values);
        Ok(())
    }

    #[test]
    fn test_hm_data_dimensions() -> TestResult {
        let h5file = hdf5::File::open(tsc())?;
        let mut hm_data = HmData::from_hdf5(&h5file)?;
        hm_data.load_full()?;

        let gps_timestamps = vec![0.0, 1.0, 2.0, 3.0];

        let (_plots, metadata) = hm_data.build_plots_and_metadata(&gps_timestamps);
        let shape = hm_data.shape();

        assert_eq!(metadata.len(), 2);

        // Verify expected shape [4, 413, 2, 76, 6, 4]
        assert_eq!(shape.len(), 6, "Should be 6-dimensional");
        assert_eq!(shape[0], 4, "First dimension should be 4");
        assert_eq!(shape[1], 413, "Second dimension should be 413");
        assert_eq!(shape[2], 2, "Third dimension should be 2");
        assert_eq!(shape[3], 76, "Fourth dimension should be 76");
        assert_eq!(shape[4], 6, "Fifth dimension should be 6");
        assert_eq!(shape[5], 4, "Sixth dimension should be 4");

        // Verify we can access boundary elements
        assert!(hm_data.get([3, 412, 1, 75, 5, 3]).is_some());
        assert!(hm_data.get([4, 0, 0, 0, 0, 0]).is_none()); // Out of bounds

        Ok(())
    }
}
