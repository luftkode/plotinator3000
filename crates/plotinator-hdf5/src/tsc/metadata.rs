use crate::util;

#[derive(Debug)]
pub(crate) struct RootMetadata {
    linear_time: f64,
    kickin_usec: f64,
    hm_ontime_usec: f64,
    gates_pr_decade_on: f64,
    metadata: Vec<(String, String)>,
}

impl RootMetadata {
    pub fn parse_from_tsc(h5: &hdf5::File) -> anyhow::Result<Self> {
        let mut metadata = vec![];
        for attr_name in h5.attr_names()? {
            let attr = h5.attr(&attr_name)?;
            let attr_str = util::read_any_attribute_to_string(&attr)?;
            metadata.push((attr_name, attr_str));
        }

        let linear_time = h5.attr("t_lin")?.read_scalar::<f64>()?;
        let kickin_usec = h5.attr("kickin_usec")?.read_scalar::<f64>()?;
        let hm_ontime_usec = h5.attr("hm_ontime_usec")?.read_scalar::<f64>()?;
        let gates_pr_decade_on = h5.attr("gates_pr_decade_on")?.read_scalar()?;

        Ok(Self {
            linear_time,
            kickin_usec,
            hm_ontime_usec,
            gates_pr_decade_on,
            metadata,
        })
    }
    /// 0-based
    pub(crate) fn last_gate_on_index(&self) -> usize {
        self.last_gate_on_count() - 1
    }
    /// 1-based (count)
    pub(crate) fn last_gate_on_count(&self) -> usize {
        let k_ramp_up = gates_pr_decade_to_k(self.gates_pr_decade_on);
        let k_ontime = gates_pr_decade_to_k(self.gates_pr_decade_on);
        let gate_kick_in = sinh_gates(self.kickin_usec, self.linear_time, k_ramp_up);

        let gate_from_kick_in_to_ramp_down: Vec<f64> = sinh_gates(
            self.hm_ontime_usec - self.kickin_usec,
            self.linear_time,
            k_ontime,
        )
        .into_iter()
        .map(|x| x + self.kickin_usec)
        .collect();

        gate_kick_in.len() + gate_from_kick_in_to_ramp_down.len()
    }

    pub(crate) fn metadata_strings(&self) -> Vec<(String, String)> {
        self.metadata.clone()
    }
}

fn gates_pr_decade_to_k(gpd: f64) -> f64 {
    1.0 / (gpd * std::f64::consts::E.log10())
}

fn sinh_gates(t: f64, t_lin: f64, k: f64) -> Vec<f64> {
    // Calculate n_max = ceil((1/k) * arcsinh(t / t_lin))
    let n_max = ((1.0 / k) * (t / t_lin).asinh()).ceil() as usize;

    // Build tgates = [t_lin * sinh(k * n) for n in 1..=n_max]
    let mut tgates: Vec<f64> = (1..=n_max)
        .map(|n| t_lin * (k * (n as f64)).sinh())
        .collect();

    // Replace last element with t
    if let Some(last) = tgates.last_mut() {
        *last = t;
    }

    tgates
}

#[cfg(test)]
mod tests {
    use plotinator_test_util::test_file_defs::tsc::tsc;
    use testresult::TestResult;

    use super::*;

    #[test]
    fn test_read_tsc_attributes() -> TestResult {
        let h5file = hdf5::File::open(tsc())?;

        let root_metadata = RootMetadata::parse_from_tsc(&h5file)?;

        let last_on = root_metadata.last_gate_on_count();
        assert_eq!(last_on, 15);

        Ok(())
    }
}
