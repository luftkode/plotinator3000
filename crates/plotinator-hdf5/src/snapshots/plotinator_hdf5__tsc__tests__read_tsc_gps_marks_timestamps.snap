---
source: crates/plotinator-hdf5/src/tsc.rs
expression: first_10_gps_time
---
[
    1.7556088924523802e18,
    1.7556088935537728e18,
    1.7556088946551662e18,
    1.7556088957565594e18,
]
